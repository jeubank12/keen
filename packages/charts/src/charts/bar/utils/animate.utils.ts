import { AnimationProps } from 'framer-motion';
import { Layout } from '@keen.io/ui-core';

import { GroupMode } from '../../../types';

type BarMotionProps = {
  x: number;
  y: number;
  height: number;
  width: number;
  groupMode: GroupMode;
  layout: Layout;
};

export const createBarMotion = ({
  groupMode,
  layout,
  height,
  width,
  x,
  y,
}: BarMotionProps): AnimationProps => {
  const motion = {
    initial: {},
    exit: {
      opacity: 0,
    },
    animate: {
      x,
      y,
      height,
      width,
      opacity: 1,
      transition: {
        duration: 0.4,
      },
    },
  };

  if (groupMode === 'grouped') {
    motion.initial = {
      opacity: 0,

      ...(layout === 'vertical'
        ? {
            x,
            y: height + y,
            height: 0,
          }
        : { y, x, width: 0 }),
    };
  } else {
    motion.initial = {
      opacity: 0,
      x,
      y,
    };
  }

  return motion;
};
