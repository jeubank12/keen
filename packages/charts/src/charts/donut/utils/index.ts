import {
  generateDonutChart,
  LabelsPosition,
  Options as DonutChartOptions,
} from './chart.utils';

export { generateDonutChart, LabelsPosition, DonutChartOptions };
