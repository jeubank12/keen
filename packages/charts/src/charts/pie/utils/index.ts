import {
  generatePieChart,
  LabelsPosition,
  Options as PieChartOptions,
} from './chart.utils';

export { generatePieChart, LabelsPosition, PieChartOptions };
