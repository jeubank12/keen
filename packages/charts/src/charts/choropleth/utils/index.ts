import { fetchMapTopology, MapType } from './map.utils';
import { generateChoropleth } from './chart.utils';

export { fetchMapTopology, generateChoropleth, MapType };
