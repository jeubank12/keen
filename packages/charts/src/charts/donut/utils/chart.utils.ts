import { sum } from 'd3-array';
import { arc, pie } from 'd3-shape';

import { calculateHypotenuse } from '../../../utils/math.utils';

import { Dimension, Margins, DataSelector } from '../../../types';

export type LabelsPosition = 'inside' | 'outside';

export type Options = {
  data: Record<string, any>[];
  labelSelector: string;
  keys: string[];
  disabledLabels: string[];
  dimension: Dimension;
  margins: Margins;
  colors: string[];
  padAngle: number;
  padRadius: number;
  cornerRadius: number;
  innerRadius: number;
  labelsRadius: number;
  labelsPosition: LabelsPosition;
};

type DonutSlice = {
  index: string;
  label: string;
  activePosition: [number, number];
  labelPosition: [number, number];
  color: string;
  selector: DataSelector;
  startAngle: number;
  endAngle: number;
};

type DonutValue = { color: string; value: number; selector: DataSelector };

export const HOVER_RADIUS = 5;

export const generateDonutChart = ({
  data,
  colors,
  dimension,
  padAngle,
  padRadius,
  cornerRadius,
  innerRadius,
  labelSelector,
  keys,
  disabledLabels,
  labelsPosition,
  labelsRadius,
  margins,
}: Options) => {
  const values: DonutValue[] = [];

  const { width, height } = dimension;
  const radius =
    Math.min(
      width - margins.left - margins.right,
      height - margins.top - margins.bottom
    ) / 2;

  data.forEach((item, idx) => {
    const label = item[labelSelector];
    if (!disabledLabels.includes(label)) {
      const result = keys.reduce((acc, currentKey) => {
        if (currentKey !== label) return acc + item[currentKey];
        return acc;
      }, 0) as number;

      values.push({
        value: result,
        selector: [idx],
        color: colors[idx],
      });
    }
  });

  const total = sum(values, d => d.value);
  const createDonut = pie().value((d: any) => d.value);

  const createArc = arc().padAngle(padAngle);

  const calculateLabelPosition = (
    startAngle: number,
    endAngle: number
  ): [number, number] => {
    const [x, y] = createArc.centroid({
      innerRadius: innerRadius,
      outerRadius: radius,
      startAngle,
      endAngle,
    });

    if (labelsPosition === 'outside') {
      return calculateHypotenuse(x, y, radius + labelsRadius);
    }

    return [x, y];
  };

  const arcs: DonutSlice[] = createDonut(values as any).map(
    ({ startAngle, endAngle, value, index, data }) => {
      const { color, selector } = data as any;
      const [x, y] = createArc.centroid({
        innerRadius: innerRadius,
        outerRadius: 0,
        startAngle,
        endAngle,
      });

      return {
        label: String(`${(Math.round(value * 100) / total).toFixed(1)}%`),
        labelPosition: calculateLabelPosition(startAngle, endAngle),
        activePosition: [x, y],
        index: String(index),
        selector,
        color,
        startAngle,
        endAngle,
      };
    }
  );

  return {
    total,
    arcs,
    drawActiveArc: arc()
      .padAngle(padAngle)
      .innerRadius(innerRadius + HOVER_RADIUS)
      .outerRadius(radius + HOVER_RADIUS)
      .padRadius(padRadius)
      .cornerRadius(cornerRadius),
    drawArc: arc()
      .padAngle(padAngle)
      .innerRadius(innerRadius)
      .outerRadius(radius)
      .padRadius(padRadius)
      .cornerRadius(cornerRadius),
  };
};
