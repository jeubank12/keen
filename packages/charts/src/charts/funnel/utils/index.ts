import { generateFunnel, calculateStepPoints } from './chart.utils';

export { generateFunnel, calculateStepPoints };
