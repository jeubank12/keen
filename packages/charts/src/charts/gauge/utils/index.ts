import { generateGauge } from './chart.utils';

export { generateGauge };
