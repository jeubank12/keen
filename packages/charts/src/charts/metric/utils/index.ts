import { generateMetric } from './metric.utils';
import { formatNumber } from './format.utils';

export { generateMetric, formatNumber };
