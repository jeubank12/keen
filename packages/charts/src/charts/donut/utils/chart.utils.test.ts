import { generateDonutChart } from './chart.utils';

import { donutChart } from '../donut-chart.fixtures';

describe('@keen/charts', () => {
  describe('<DonutChart /> - utils', () => {
    it('should create arcs based on provided data', () => {
      const { arcs } = generateDonutChart(donutChart);
      const { data } = donutChart;

      expect(arcs.length).toEqual(data.length);
    });

    it('should calculate properties for each arc', () => {
      const { arcs } = generateDonutChart(donutChart);

      expect(arcs).toMatchSnapshot();
    });

    it('should calculate position for labels placed "outer" of slice', () => {
      const { arcs } = generateDonutChart({
        ...donutChart,
        labelsPosition: 'outside',
      });

      expect(arcs).toMatchSnapshot();
    });
  });
});
