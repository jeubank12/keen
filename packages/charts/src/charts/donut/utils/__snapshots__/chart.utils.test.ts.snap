// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen/charts <DonutChart /> - utils should calculate position for labels placed "outer" of slice 1`] = `
Array [
  Object {
    "activePosition": Array [
      -4.312731892445443,
      -9.022213898145175,
    ],
    "color": "#85B4C3",
    "endAngle": 6.283185307179586,
    "index": "4",
    "label": "14.2%",
    "labelPosition": Array [
      -30.1891232471181,
      -63.155497287016225,
    ],
    "selector": Array [
      0,
    ],
    "startAngle": 5.391378360354096,
  },
  Object {
    "activePosition": Array [
      -5.545165057524355,
      8.321727253690241,
    ],
    "color": "#CB5623",
    "endAngle": 4.377961375325131,
    "index": "2",
    "label": "20.6%",
    "labelPosition": Array [
      -38.81615540267048,
      58.25209077583169,
    ],
    "selector": Array [
      1,
    ],
    "startAngle": 3.080787634488055,
  },
  Object {
    "activePosition": Array [
      7.106806583781534,
      -7.03514748819948,
    ],
    "color": "#E29B1E",
    "endAngle": 1.5809304966451863,
    "index": "0",
    "label": "25.2%",
    "labelPosition": Array [
      49.74764608647073,
      -49.24603241739636,
    ],
    "selector": Array [
      2,
    ],
    "startAngle": 0,
  },
  Object {
    "activePosition": Array [
      -9.851963176211386,
      -1.7142991496745383,
    ],
    "color": "#487650",
    "endAngle": 5.391378360354096,
    "index": "3",
    "label": "16.1%",
    "labelPosition": Array [
      -68.9637422334797,
      -12.000094047721769,
    ],
    "selector": Array [
      3,
    ],
    "startAngle": 4.377961375325131,
  },
  Object {
    "activePosition": Array [
      7.247927872291199,
      6.889669190756866,
    ],
    "color": "#F4A083",
    "endAngle": 3.080787634488055,
    "index": "1",
    "label": "23.9%",
    "labelPosition": Array [
      50.7354951060384,
      48.227684335298065,
    ],
    "selector": Array [
      4,
    ],
    "startAngle": 1.5809304966451863,
  },
]
`;

exports[`@keen/charts <DonutChart /> - utils should calculate properties for each arc 1`] = `
Array [
  Object {
    "activePosition": Array [
      -4.312731892445443,
      -9.022213898145175,
    ],
    "color": "#85B4C3",
    "endAngle": 6.283185307179586,
    "index": "4",
    "label": "14.2%",
    "labelPosition": Array [
      -12.938195677336328,
      -27.066641694435525,
    ],
    "selector": Array [
      0,
    ],
    "startAngle": 5.391378360354096,
  },
  Object {
    "activePosition": Array [
      -5.545165057524355,
      8.321727253690241,
    ],
    "color": "#CB5623",
    "endAngle": 4.377961375325131,
    "index": "2",
    "label": "20.6%",
    "labelPosition": Array [
      -16.635495172573066,
      24.965181761070724,
    ],
    "selector": Array [
      1,
    ],
    "startAngle": 3.080787634488055,
  },
  Object {
    "activePosition": Array [
      7.106806583781534,
      -7.03514748819948,
    ],
    "color": "#E29B1E",
    "endAngle": 1.5809304966451863,
    "index": "0",
    "label": "25.2%",
    "labelPosition": Array [
      21.320419751344602,
      -21.10544246459844,
    ],
    "selector": Array [
      2,
    ],
    "startAngle": 0,
  },
  Object {
    "activePosition": Array [
      -9.851963176211386,
      -1.7142991496745383,
    ],
    "color": "#487650",
    "endAngle": 5.391378360354096,
    "index": "3",
    "label": "16.1%",
    "labelPosition": Array [
      -29.55588952863416,
      -5.142897449023615,
    ],
    "selector": Array [
      3,
    ],
    "startAngle": 4.377961375325131,
  },
  Object {
    "activePosition": Array [
      7.247927872291199,
      6.889669190756866,
    ],
    "color": "#F4A083",
    "endAngle": 3.080787634488055,
    "index": "1",
    "label": "23.9%",
    "labelPosition": Array [
      21.743783616873596,
      20.669007572270598,
    ],
    "selector": Array [
      4,
    ],
    "startAngle": 1.5809304966451863,
  },
]
`;
